src/main/java/org/folio/tlib/postgres/impl/PgCqlQueryImpl.java

```diff
@@ -0,0 +1,270 @@
+package org.folio.tlib.postgres.impl;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.tlib.postgres.PgCqlField;
+import org.folio.tlib.postgres.PgCqlQuery;
+import org.z3950.zing.cql.CQLBooleanNode;
+import org.z3950.zing.cql.CQLNode;
+import org.z3950.zing.cql.CQLParseException;
+import org.z3950.zing.cql.CQLParser;
+import org.z3950.zing.cql.CQLPrefixNode;
+import org.z3950.zing.cql.CQLSortNode;
+import org.z3950.zing.cql.CQLTermNode;
+
+public class PgCqlQueryImpl implements PgCqlQuery {
+  private static final Logger log = LogManager.getLogger(PgCqlQueryImpl.class);
+
+  final CQLParser parser = new CQLParser(CQLParser.V1POINT2);
+  final Map<String, PgCqlField> fields = new HashMap<>();
+
+  String language = "english";
+  CQLNode cqlNodeRoot;
+
+  @Override
+  public void parse(String query) {
+    if (query == null) {
+      cqlNodeRoot = null;
+    } else {
+      try {
+        log.debug("Parsing {}", query);
+        cqlNodeRoot = parser.parse(query);
+      } catch (CQLParseException | IOException e) {
+        throw new IllegalArgumentException(e.getMessage());
+      }
+    }
+  }
+
+  @Override
+  public String getWhereClause() {
+    return whereHandle(cqlNodeRoot);
+  }
+
+  static String basicOp(CQLTermNode termNode) {
+    String base = termNode.getRelation().getBase();
+    switch (base) {
+      case "==":
+        return "=";
+      case "=":
+      case "<>":
+        return base;
+      default:
+        throw new IllegalArgumentException("Unsupported operator " + base + " for: "
+            + termNode.toCQL());
+    }
+  }
+
+  static String numberOp(CQLTermNode termNode) {
+    String base = termNode.getRelation().getBase();
+    switch (base) {
+      case "==":
+        return "=";
+      case "=":
+      case "<>":
+      case ">":
+      case "<":
+      case "<=":
+      case ">=":
+        return base;
+      default:
+        throw new IllegalArgumentException("Unsupported operator " + base + " for: "
+            + termNode.toCQL());
+    }
+  }
+
+  /**
+   * See if this is a CQL query with a existence check (NULL or NOT NULL).
+   * <p>Empty term makes "IS NULL" for CQL relation =, "IS NOT NULL" for CQL relation <>.
+   * </p>
+   * @param field CQL field.
+   * @param termNode term.
+   * @return SQL op for NULL; null if not a NULL check.
+   */
+  String handleNull(PgCqlField field, CQLTermNode termNode) {
+    if (!termNode.getTerm().isEmpty()) {
+      return null;
+    }
+    String base = termNode.getRelation().getBase();
+    switch (base) {
+      case "==":
+        return null;
```

Comment:
Nice!

Special cases like this are going to need documenting, though.

Reply:
Yeah, that empty string is a special case. Not sure it's so useful, but in a UI, when leaving a term as empty (don't care).. The TRUE outcome comes in natural (and as you have seen everywhere. return null from WHERE means TRUE.. So WHERE can be omitted if that what it ends up being).